renderc/test/render_test.cc

```diff
@@ -0,0 +1,297 @@
+  ASSERT_EQ(handle.access(index).number(), the_number);
+
+  handle.put(index);
+  ASSERT_FALSE(handle.valid());
+
+  ASSERT_EQ(index.simple_span.size(), 0);
+}
+
+TEST(RenderTest, IndexedSpan)
+{
+  static constexpr u32 key = 42;
+
+  test::app1::Index index;
+
+  {
+    auto ahandle = index.indexed_span.by_key(key);
+    ASSERT_TRUE(ahandle.valid());
+    ASSERT_EQ(ahandle.number(), key);
+
+    ASSERT_EQ(index.indexed_span.size(), 1);
+
+    {
+      auto another = index.indexed_span.by_key(key);
+      ASSERT_TRUE(another.valid());
+
+      // Still only one span is allocated.
+      ASSERT_EQ(index.indexed_span.size(), 1);
+
+      // It's the same span.
+      ASSERT_EQ(ahandle.loc(), another.loc());
+    }
+
+    {
+      auto different = index.indexed_span.by_key(key + 1);
+      ASSERT_TRUE(different.valid());
+
+      // Additional span has been allocated.
+      ASSERT_EQ(index.indexed_span.size(), 2);
+
+      // It's not the same span.
+      ASSERT_NE(ahandle.loc(), different.loc());
+    }
+  }
+
+  ASSERT_EQ(index.indexed_span.size(), 0);
+}
+
+TEST(RenderTest, MetricStore)
+{
+  using metrics_visitor_t =
+      std::function<void(u64, test::app1::weak_refs::metrics_span, test::metrics::some_metrics const &, u64)>;
+
+  static constexpr u64 timeslot_duration = 1'000'000'000;
+  u64 time_now = 1;
+
+  test::app1::Index index;
+
+  {
+    auto span = index.metrics_span.alloc();
+    ASSERT_TRUE(span.valid());
+
+    ASSERT_EQ(index.metrics_span.size(), 1);
+    ASSERT_EQ(span.refcount(), 1);
+
+    test::metrics::some_metrics_point metrics{
+        .active = 55,
+        .total = 100,
+    };
+    span.metrics_update(time_now, metrics);
+
+    ASSERT_EQ(index.metrics_span.size(), 1);
+
+    // Metric store is keeping a reference to this span.
+    ASSERT_EQ(span.refcount(), 2);
+  }
+
+  // Metric store is keeping the span allocated.
+  ASSERT_EQ(index.metrics_span.size(), 1);
+
+  // Advance the current time.
+  time_now += 2 * timeslot_duration;
+
+  // Metrics slot should be ready.
+  ASSERT_TRUE(index.metrics_span.metrics_ready(time_now));
+
+  // Get the metrics from the current slot.
+  int metric_counter{0};
+  metrics_visitor_t on_metric = [&metric_counter](u64 timestamp, auto span, auto metrics, u64 interval) { ++metric_counter; };
+  index.metrics_span.metrics_foreach(time_now, on_metric);
+
+  // Only one metrics slot.
+  ASSERT_EQ(metric_counter, 1);
+
+  // Metrics store should be cleared out.
+  ASSERT_TRUE(index.metrics_span.metrics.current_queue().empty());
+
+  // Metric store should no longer keep a reference to the span.
+  ASSERT_EQ(index.metrics_span.size(), 0);
+}
+
+TEST(RenderTest, ManualReference)
+{
+  test::app1::Index index;
+
+  auto span = index.span_with_manual_reference.alloc();
+  ASSERT_TRUE(span.valid());
+
+  auto simple_loc = span.manual_reference().loc();
+
+  // Currently the reference is an invalid reference.
+  ASSERT_EQ(simple_loc, span.manual_reference().invalid);
```

Comment:
nit, for future proofing consider not relying that `auto_handle.valid()` is equivalent to `auto_handle.loc() != auto_handle.invalid` (so the internal implementation could change without breaking the test)

Reply:
This test is not relying on that. It is relying on `handle.loc() == handle.invalid` being sufficient condition that the handle is invalid. Logical negation of that statement makes `handle.loc() != handle.invalid` a necessary condition for validity, but that is not what the test is doing.

Reply:
Agree. Was just sayin that even though the const ‘invalid’ is ‘public’, we might still want to consider it an internal implementation detail in this test to make the test less fragile. (as mentioned it’s a nit comment, more following up our offline discussion on test fragility)